src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs

```diff
@@ -0,0 +1,334 @@
+                // Even though this is already handled by JsonMetadataServicesConverter,
+                // this avoids creating a WriteStack and calling into the converter infrastructure.
+
+                Debug.Assert(SerializeHandler != null);
+                Debug.Assert(Options.SerializerContext?.CanUseSerializationLogic == true);
+                Debug.Assert(Converter is JsonMetadataServicesConverter<T>);
+
+                SerializeHandler(writer, rootValue!);
+                writer.Flush();
+            }
+            else if (
+#if NETCOREAPP
+                !typeof(T).IsValueType &&
+#endif
+                Converter.CanBePolymorphic &&
+                rootValue is not null &&
+                Options.TryGetPolymorphicTypeInfoForRootType(rootValue, out JsonTypeInfo? derivedTypeInfo))
+            {
+                Debug.Assert(typeof(T) == typeof(object));
+                derivedTypeInfo.SerializeAsObject(writer, rootValue, isInvokedByPolymorphicConverter: true);
+                // NB flushing is handled by the derived type's serialization method.
+            }
+            else
+            {
+                WriteStack state = default;
+                state.Initialize(this, rootValueBoxed, isPolymorphicRootValue: isInvokedByPolymorphicConverter);
+
+                bool success = EffectiveConverter.WriteCore(writer, rootValue!, Options, ref state);
+                Debug.Assert(success);
+                writer.Flush();
+            }
+        }
+
+        // Root serialization method for async streaming serialization.
+        internal async Task SerializeAsync(
+            Stream utf8Json,
+            T? rootValue,
+            CancellationToken cancellationToken,
+            object? rootValueBoxed = null,
+            bool isInvokedByPolymorphicConverter = false)
+        {
+            Debug.Assert(IsConfigured);
+            Debug.Assert(rootValueBoxed is null || rootValueBoxed is T);
+
+            if (CanUseSerializeHandlerInStreaming)
+            {
+                // Short-circuit calls into SerializeHandler, if the `CanUseSerializeHandlerInStreaming` heuristic allows it.
+
+                Debug.Assert(SerializeHandler != null);
+                Debug.Assert(Options.SerializerContext?.CanUseSerializationLogic == true);
+                Debug.Assert(Converter is JsonMetadataServicesConverter<T>);
+
+                using var bufferWriter = new PooledByteBufferWriter(Options.DefaultBufferSize);
+                using var writer = new Utf8JsonWriter(bufferWriter, Options.GetWriterOptions());
+
+                try
+                {
+                    SerializeHandler(writer, rootValue!);
+                    writer.Flush();
+                    await bufferWriter.WriteToStreamAsync(utf8Json, cancellationToken).ConfigureAwait(false);
+                }
+                finally
+                {
+                    OnRootLevelAsyncSerializationCompleted(writer.BytesCommitted + writer.BytesPending);
```

Comment:
Yes. Moving serialization to the non-streaming fast path admits the possibility of it failing with OOM's, so we want to be able to invalidate the optimization as soon as it happens.

Reply:
This comment is on the wrong line - it should be for the `else` case, which gets called the first 10 times.

My thinking is that if an exception is thrown, the full payload hasn't been written yet. If we call `OnRootLevelAsyncSerializationCompleted` with less than the full payload of bytes, it could count towards the heuristic, when it really shouldn't.